src/qibo/core/states.py

```diff
@@ -92,7 +92,7 @@ def wrapper(self, qubits=None, measurement_gate=None):
     def probabilities(self, qubits=None, measurement_gate=None):
         unmeasured_qubits = tuple(i for i in range(self.nqubits)
                                   if i not in qubits)
-        state = K.reshape(K.square(K.abs(K.cast(self.tensor))), self.nqubits * (2,))
+        state = K.reshape(K.square(K.abs(K.cast(self.tensor, dtype="complex128"))), self.nqubits * (2,))
```

Comment:
Does it work if we make the cast on the float level, after the abs:
```suggestion
        state = K.reshape(K.cast(K.square(K.abs(self.tensor)), dtype="float64"), self.nqubits * (2,))
```
This way we would save some memory and performance deterioration, right? Not sure if it works though.

Reply:
I'm not sure, we may still need the cast because otherwise we need to cast tensor from cupy to numpy in case of fallback, but we may add an additional cast before the sum. I'll give it a try.

Reply:
I tried a similar approach and it works, take a look at then latest commit.

Reply:
Yes, that's exactly what I meant. Given that `K.cast` is properly implemented and does not create any copies when the same type is passed, now it should work ok, right? And the sum issue is fixed since we convert to float64 before, while we avoid the complex128 conversion.

Reply:
Yes, it should work fine.